src/main/java/org/openstreetmap/atlas/geography/atlas/items/complex/roundabout/ComplexRoundabout.java

```diff
@@ -0,0 +1,295 @@
+
+        // The cross product tells us the direction of the orthogonal vector, which is
+        // Directly related to the direction of rotation/traffic
+        return (vector1X * vector2Y) - (vector1Y * vector2X);
+    }
+
+    public ComplexRoundabout(final Edge source)
+    {
+        this(source, Arrays.asList("AIA", "ATG", "AUS", "BGD", "BHS", "BMU", "BRB", "BRN", "BTN",
+                "BWA", "CCK", "COK", "CXR", "CYM", "CYP", "DMA", "FJI", "FLK", "GBR", "GGY", "GRD",
+                "GUY", "HKG", "IDN", "IMN", "IND", "IRL", "JAM", "JEY", "JPN", "KEN", "KIR", "KNA",
+                "LCA", "LKA", "LSO", "MAC", "MDV", "MLT", "MOZ", "MSR", "MUS", "MWI", "MYS", "NAM",
+                "NFK", "NIU", "NPL", "NRU", "NZL", "PAK", "PCN", "PNG", "SGP", "SGS", "SHN", "SLB",
+                "SUR", "SWZ", "SYC", "TCA", "THA", "TKL", "TLS", "TON", "TTO", "TUV", "TZA", "UGA",
+                "VCT", "VGB", "VIR", "WSM", "ZAF", "ZMB", "ZWE"));
+    }
+
+    public ComplexRoundabout(final Edge source, final List leftDrivingCountries)
+    {
+        super(source);
+        try
+        {
+            if (!(
+            // Make sure that the source has an iso_country_code
+            source.getTag(ISOCountryTag.KEY).isPresent()
+                    // Make sure that the source is an instances of a roundabout
+                    && JunctionTag.isRoundabout(source)
+                    // Make sure that we are looking at a master edge
+                    && (source.isMasterEdge())))
+            {
+                throw new CoreException("Invalid source edge for a roundabout.");
+            }
+            final String isoCountryCode = source.tag(ISOCountryTag.KEY).toUpperCase();
+
+            // Get all edges in the roundabout
+            this.roundaboutEdgeSet = new SimpleEdgeWalker(source, this.isRoundaboutEdge())
+                    .collectEdges();
+
+            // Try to build a Route from the edges
+            try
+            {
+                this.roundaboutRoute = Route.fromNonArrangedEdgeSet(this.roundaboutEdgeSet, false);
+            }
+            // If a Route cannot be formed, flag the edges as an incomplete roundabout.
+            catch (final CoreException badRoundabout)
+            {
+                this.invalidationReasons.add(INCOMPLETE_ROUTE_INSTRUCTIONS);
+                throw new CoreException(INCOMPLETE_ROUTE_INSTRUCTIONS);
+            }
+
+            // Flag if any of the edges are not car navigable or master Edges, or the route does not
+            // form a closed loop.
+            if (this.roundaboutEdgeSet.stream()
+                    .anyMatch(roundaboutEdge -> !HighwayTag.isCarNavigableHighway(roundaboutEdge)
+                            || !roundaboutEdge.isMasterEdge())
+                    || !this.roundaboutRoute.start().inEdges().contains(this.roundaboutRoute.end()))
+            {
+                this.invalidationReasons.add(INCOMPLETE_ROUTE_INSTRUCTIONS);
```

Comment:
Instead of storing our own invalid reasons, should we just use the parent class's `setInvalidReason` list function?

Reply:
I am actually using both. The reason for this is the parent `setInvalidReason` only stores one reason, and we need to store multiple to pass to atlas checks.